semimpl/do.go

```diff
@@ -0,0 +1,73 @@
+package semimpl
+
+import (
+	"context"
+	"errors"
+	"log"
+	"sync"
+
+	"golang.org/x/sync/semaphore"
+)
+
+// Do concurrently executes callback at most maxIter times or until ctx is done
+// or canceled. Concurrency is handled leveraging the semaphore pattern, which
+// ensures at most numWorkers goroutines are spawned at the same time.
+func Do(ctx context.Context, numWorkers, maxIter int, callback func()) {
+	numWorkers = sanitizeNumWorkers(numWorkers)
+	maxIter = sanitizeMaxIter(maxIter)
+	callback = sanitizeCallback(callback)
+
+	sem := semaphore.NewWeighted(int64(numWorkers))
+	wg := sync.WaitGroup{}
+
+	for i := 0; i < maxIter || maxIter == 0; i++ {
+		wg.Add(1)
+
+		if err := sem.Acquire(ctx, 1); err != nil {
+			handleContextError(err)
+			wg.Done()
+			break
+		}
+
+		go func() {
+			defer func() {
+				sem.Release(1)
+				wg.Done()
+			}()
+			callback()
+		}()
+	}
+
+	wg.Wait()
+}
+
+func sanitizeNumWorkers(numWorkers int) int {
+	if numWorkers < 1 {
+		return 1
+	}
+	return numWorkers
+}
+
+func sanitizeMaxIter(maxIter int) int {
+	if maxIter < 0 {
+		return 0
+	}
+	return maxIter
+}
+
+func sanitizeCallback(callback func()) func() {
+	if callback == nil {
+		return func() {}
+	}
+	return callback
+}
+
+func handleContextError(err error) {
```

Comment:
We could just do nothing: it's not _really_ an error as it won't happen unexpectedly, it's just how a `context` tells us it has timeout / been canceled, and we have full control over it on its creation.
The real error handling here is `default: log.Fatal(err)` (to be improved later), the rest are non-errors.

Reply:
Oh, maybe you meant what do we do instead of `log.Fatal`?
Well I just checked, this case should never happen: the only error values ever returned by `sem.Acquire` are `nil` and the `context` errors which are handled here.